# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Lane 4: CTR / Engagement Opportunity Scoring

I feel like lane fourth is the most efficient way to study quant finance. CTR adjusted for position and volume, opportunity scoring with filters, working around statistical noise from low-volume pages. It's signal-precise and metric-heavy, similar to quant finance work in real life.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Which visible pages under-capture clicks or engagement relative to their position and impression volume?

The action someone could take: A content reviewer examines the top-ranked pages and rewrites title tags, meta descriptions, or on-page content to better match searchers’ intentions and improve CTR.

The cost of a wrong call: The content team wastes time on pages that don't need fixing, while genuinely underperforming pages go unreviewed. This could reduce the effectiveness of already well-performing pages also.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [42]:
import pandas as pd
import numpy as np

df = pd.read_csv("content_refresh_anonymized.csv")
zero_df = df.loc[(df['impressions_90d']>=100) & (df['clicks_90d']==0)]
print('The number of pages with impressions larger than 100, but zero clicks: ',zero_df.shape[0], '\n')
print(f"That was around 20% of the dataset (of {df.shape[0]} pages)\n")
print('That statistics for each position_tier in the group: \n',zero_df['position_tier'].value_counts())
print('\n It can be seen that within the 6068 pages that have 100+ impressions, no clicks, around 20% of them are page_1 position_tier.\n')
print('This is clearly a problem, suggesting that this data is worth my next 7 weeks')


The number of pages with impressions larger than 100, but zero clicks:  6068 

That was around 20% of the dataset (of 30000 pages)

That statistics for each position_tier in the group: 
 position_tier
page_3_5    2331
striking    1593
page_1      1347
deep         690
top_3        107
Name: count, dtype: int64

 It can be seen that within the 6068 pages that have 100+ impressions, no clicks, around 20% of them are page_1 position_tier.

This is clearly a problem, suggesting that this data is worth my next 7 weeks


Pages at position_tier page_1 should be getting clicks (given they already have a fair amount of impressions).

In [34]:
# Filter to pages with meaningful volume (100+ impressions)
substantial = df[df['impressions_90d'] >= 100]

# Group by position tier and look at CTR stats
substantial.groupby('position_tier')['ctr'].describe()

                count      mean       std  min   25%   50%   75%    max
position_tier                                                          
deep            879.0  0.055415  0.169861  0.0  0.00  0.00  0.00   2.08
page_1         8633.0  0.354760  0.502342  0.0  0.09  0.23  0.46  11.76
page_3_5       6058.0  0.142359  0.229041  0.0  0.00  0.06  0.19   3.38
striking       5903.0  0.255782  0.346695  0.0  0.00  0.15  0.34   5.43
top_3           533.0  0.334128  0.487466  0.0  0.05  0.19  0.48   6.19


It is clear that the page_1 pages averages at ctr = 0.35% for 100+ impressions
-> Any page_1 pages with 100+ impressions but have significantly lower CTR are losing opportunity and are underperforming.

What can be concluded is CRT varies dramatically by position_tier (when page_1 averages at 0.35% but page_3_5 averages at 0.14%). Therefor pages at position_tier 1 that aren't getting clicks even with 100+ impressions represent opportunity.


In [50]:
df = pd.read_csv("content_refresh_anonymized.csv")
#df.describe()[['impressions_90d','ctr']]
#df[['position_tier','impressions_90d']].head(100)
zero_df = df.loc[(df['impressions_90d']>=250) & (df['clicks_90d']==0)]
print('Changing the impressions_90d threshold to be 250+, the number of pages with 0 clicks are: ', zero_df.shape[0],'\n')
print('The value counts of each position tier within that group: \n')
zero_df['position_tier'].value_counts()

print(f'The percentage of pages at tier 1 getting no clicks with 250+ impressions: {((851/4166)*100):.3f}%, the pattern repeats even at a higher threshold.')

Changing the impressions_90d threshold to be 250+, the number of pages with 0 clicks are:  4166 

The value counts of each position tier within that group: 

The percentage of pages at tier 1 getting no clicks with 250+ impressions: 20.427%


In [45]:
substantial = df[df['impressions_90d']>250]

substantial.groupby('position_tier')['ctr'].describe()

,count,mean,std,min,25%,50%,75%,max
position_tier,,,,,,,,
deep,618.0,0.044515,0.134874,0.0,0.00,0.00,0.01,2.08
page_1,7807.0,0.336804,0.371476,0.0,0.11,0.23,0.44,5.42
page_3_5,5220.0,0.141123,0.202115,0.0,0.00,0.08,0.20,3.38
striking,5232.0,0.257955,0.318999,0.0,0.05,0.17,0.34,5.43
top_3,495.0,0.329051,0.417329,0.0,0.06,0.19,0.48,5.19


This pattern can also be seen at the 250+ thresholds of impressions_90d, when page_1 have higher CRT (at 0.34%) than page_3_5 (at 0.14%).

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

Results are observational. Low CTR is associated with potential title / meta issues but does not immediately prove causation. Low CTR may also reflect intent mismatch, SERP feature competition, or seasonal patterns.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.